### PR TITLE
[add] fast TDM augmentation

### PR DESCRIPTION
**Pull request by lsg1213**

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -158,33 +158,35 @@ def get_both_dataset(config, mode:str='train'):
...
-    x = [get_preprocessed_x(x_, sr, mode=mode,
+    TDM_PATH = './'
+    tdm_x, tdm_y = get_TDMset(TDM_PATH)
+    x, y = TDM_aug(x, y, tdm_x, tdm_y, 
+                   max_overlap_num=max_overlap_num, 
+                   max_overlap_per_frame=max_overlap_per_frame, 
+                   min_overlap_sec=min_overlap_sec, 
+                   max_overlap_sec=max_overlap_sec)
+    del tdm_x, tdm_y
+    x = list(map(lambda x_: get_preprocessed_x_tf(x_, sr, mode=mode,
```
Comment (daniel03c1) [inline train.py:181]: 여기 dataset을 사용하지 않더라도, 여러가지 multiprocessing 라이브러리 이용해서라도 병렬화하면
어떨까요?

File: data_loader.py [hunk 2]
```diff
@@ -122,12 +123,11 @@ def preprocess_label(labels, max_label_length=600):
...
 
         if cur_len < max_len: 
-            labels = np.pad(labels, ((0, max_len-cur_len), (0,0)), 'constant')
+            labels = tf.pad(labels, ((0, max_len-cur_len), (0,0)))
         else:
             labels = labels[:max_len]
         return labels
-    
-    x = list(map(lambda x: torchaudio.load(x)[0], f_paths))
+    x = list(map(lambda x: tf.transpose(tf.audio.decode_wav(tf.io.read_file(x))[0]), f_paths))
```
Comment (daniel03c1) [inline data_loader.py:130]: 이 파트가 들어간 함수 이름이 get_preprocessed_wave인데, 
첫번째는 전처리 하지 않은 그냥 단순히 load인점, 두번째는 wave 뿐만 아니라 레이블도 같이 출력된다는 점에서
헷깔려서 이름을 바꾸는게 좋을것 같은데 어떠신가요?

File: data_loader.py [hunk 3]
```diff
@@ -174,76 +174,99 @@ def get_TDMset(TDM_PATH):
...
+def TDM_aug(x: list, y: list, tdm_x, tdm_y, sr=24000, label_resolution=0.1, max_overlap_num=5, max_overlap_per_frame=2, min_overlap_sec=1, max_overlap_sec=5):
     '''
-        x: list(torch.Tensor): shape(sample number, channel(4), frame(1440000))
-        y: list(np.ndarray): shape(sample number, time(600), class+cartesian(14+42))
-        tdm_x: list(np.ndarray): shape(class_num, channel(4), frame)
-        tdm_y: list(np.ndarray): shape(class_num, time, class+cartesian(14+42))
+        x: list(tf.Tensor): shape(sample number, channel(4), frame(1440000))
+        y: list(tf.Tensor): shape(sample number, time(600), class+cartesian(14+42))
+        tdm_x: list(tf.Tensor): shape(class_num, channel(4), frame)
+        tdm_y: list(tf.Tensor): shape(class_num, time, class+cartesian(14+42))
```
Comment (daniel03c1) [inline data_loader.py:195]: sample number가 number of samples 말하는거같은데, 
몇번째 샘플인지 인덱스를 의미하는 것처럼 보이는데, 주석에 sample number를 바꾸면 어떨까요?
마찬가지로 class_num이 몇번째 클래스인지로 해석되는것 같습니다.
그리고 혹시 주석을 보면 3차원 데이터로 보이는데, 원래는 list여서 길이가 다를 수 있는거죠?

File: data_loader.py [hunk 4]
```diff
@@ -174,76 +174,99 @@ def get_TDMset(TDM_PATH):
...
     '''
     class_num = y[0].shape[-1] // 4
-    min_overlap_time *= int(1 / label_resolution) #  
-    max_overlap_time *= int(1 / label_resolution) # 
+    min_overlap_sec = int(min_overlap_sec / label_resolution) 
+    max_overlap_sec = int(max_overlap_sec / label_resolution)
```
Comment (daniel03c1) [inline data_loader.py:199]: 이 파트에 여러번 리뷰달아 죄송합니다. 함수 입력으로 받을 때는 초 단위니까 sec이긴 한데,
label_resolution으로 나누면 초단위가 아니라 프레임단위인거죠? 다른 역할이니까 두개 다른 이름을 주는건 어떤가요?

File: data_loader.py (hunk 3 above)
Comment (lsg1213) [reply data_loader.py:195]: tdm은 각 class 마다 길이가 달라서 저 상태로 유지해야해요 그리고 다른 말씀은 듣고보니 공감되네요 바꾸겠습니다

File: data_loader.py (hunk 4 above)
Comment (lsg1213) [reply data_loader.py:199]: 약간 이름 아이디어가 좀 애매한데 추천해주실만한 이름이 있을까요?

File: data_loader.py (hunk 2 above)
Comment (lsg1213) [reply data_loader.py:130]: 혹시 어떤 이름이 좋을까요? get_preprocessed_data 이런식으로 할까요?
처음에 저 이름으로 지은 이유는 wave 이지만 padding이라는 preprocess를 해주기 때문에 저런 이름을 넣었거든요 wave 상태에서 해줘야하는 모든 전처리는 padding 밖에 없어서요

File: train.py (hunk 1 above)
Comment (lsg1213) [reply train.py:181]: 저도 concurrent로 해볼랬는데 이게 process쪽이 잘 안 되네요 말씀대로 multiprocessing을 알아보겠습니다 아 그리고 get_preprocessed_x_tf 는 현재 15초 걸려서 길지는 않긴 합니다

Comment (lsg1213) [reply train.py:181]: 뒤에 dataloader 다 포함해서 15초입니다

File: data_loader.py [hunk 5]
```diff
@@ -174,76 +174,99 @@ def get_TDMset(TDM_PATH):
...
-    
-    weight = 1 / torch.tensor([i.shape[0] for i in tdm_y])
-    weight /= weight.sum()
-    weight = weight.cumsum(-1)
+
     def add_noise(i):
-        selected_cls = weight.multinomial(max_overlap_num, replacement=True) # (max_overlap_num,)
+        weight = 1 / tf.convert_to_tensor([k.shape[0] for k in tdm_y])
+        weight /= tf.reduce_sum(weight)
+        selected_cls = tf.random.categorical(tf.math.log(weight[tf.newaxis,...]), max_overlap_num)[0] # (max_overlap_num,)
```
Comment (daniel03c1) [inline data_loader.py:205]: 길이의 역수로 하면 오히려 분포가 uniform하지 않게 되고, 노이즈가 더 많이 합쳐질수록 타겟으로 하는 uniform과는 거리가
멀어지지 않나요?
어짜피 클래스별로 이미 나눠놨기 때문에 그 중에서 uniform하게 샘플링하면 합치는 노이즈가 많아질수록 uniform으로 수렴하지 않을까요?
물론 샘플수가 적은 상태에서 밸런스를 최대한 맞추는 방법은 매번 노이즈가 추가되기에, 미리 만들어놓은 tdm_y가 아니라,
합쳐지고 있는 y[i]를 기준으로 밸런스를 맞출 수 있을 것 같은데 이 또한 역수가 답이 되지 않을 것 같습니다.
보통 역수는 sampling은 건들지 않고 샘플당 weight만 조절할 때 사용하는데, 분포에는 적절하지 않은 것 같은 생각이 듭니다.

File: data_loader.py [hunk 6]
```diff
@@ -174,76 +174,99 @@ def get_TDMset(TDM_PATH):
...
+
+        def _add_noise(cls):
```
Comment (daniel03c1) [inline data_loader.py]: 다시 예전처럼 함수의 입력으로 외부 데이터를 함수 내부에서 변경하는 함수로 돌아간 이유가
메모리 leak을 해결못해서 이렇게 구현하신건가요?

File: data_loader.py [hunk 7]
```diff
@@ -174,76 +174,99 @@ def get_TDMset(TDM_PATH):
...
+        cond = lambda i, j: j < len(selected_cls)
+        def body(i, j):
+            _add_noise(selected_cls[j])
+            return i, j + 1
+        tf.while_loop(cond, body, (i, j))
+        return tf.zeros((), dtype=tf.int32)
+
+
+    tf.map_fn(add_noise, tf.range(len(x)))
+    return x, y
```
Comment (daniel03c1) [inline data_loader.py]: 사소하긴한데, 함수 안에서는 한줄씩만 띄고 함수 밖에서는 두줄씩 띄면 더 보기 좋을 것 같습니다!

File: data_loader.py (hunk 6 above)
Comment (daniel03c1) [inline data_loader.py]: 혹시 몇번쨰 클래스를 합성할지 함수의 입력으로 넣은 이유가 있을까요?
단순히 랜덤한 노이즈를 합성하기 위해서는 몇번째 클래스를 합성할지 입력받을 필요 없이
함수 내부에서 랜덤하게 클래스를 선택하면 되는 것 같아서요.

File: data_loader.py [hunk 8]
```diff
@@ -174,76 +174,99 @@ def get_TDMset(TDM_PATH):
...
     def add_noise(i):
-        selected_cls = weight.multinomial(max_overlap_num, replacement=True) # (max_overlap_num,)
+        weight = 1 / tf.convert_to_tensor([k.shape[0] for k in tdm_y])
+        weight /= tf.reduce_sum(weight)
+        selected_cls = tf.random.categorical(tf.math.log(weight[tf.newaxis,...]), max_overlap_num)[0] # (max_overlap_num,)
+
+        def _add_noise(cls):
+            frame_y_num = y[i].shape[0]
+            sample_time = tf.random.uniform((), min_overlap_sec, max_overlap_sec,dtype=tf.int64) # to milli second
+            offset = tf.random.uniform((), 0, frame_y_num - sample_time, dtype=tf.int64) # offset as label
```
Comment (daniel03c1) [inline data_loader.py:210]: 위에 sample number 파트랑 비슷한 상황인데, frame_y_num도 마찬가지로 frame_y의 number로 읽히는데
보니까 index가 아니라 몇개있는지이니까, num을 앞으로 옮기면 어떨까요?

File: train.py (hunk 1 above)
Comment (daniel03c1) [reply train.py:181]: 나중에 이것저것 여러가지 테스트하고, 자주 tdm 적용하려면 최대한 시간 줄이면 좋을 것 같아서요! 알겠습니다!

File: data_loader.py (hunk 2 above)
Comment (daniel03c1) [reply data_loader.py:130]: 음... 바로 의미 알 수 있게 load_wav_and_label 이런식은 어떤가요?

File: data_loader.py (hunk 4 above)
Comment (daniel03c1) [reply data_loader.py:199]: 음 제일 간단하게는 min_overlap_frame = int(min_overlap_sec / ...) 이런식으로 하는건 어떤가요?

File: data_loader.py (hunk 2 above)
Comment (lsg1213) [reply data_loader.py:130]: 변경했습니다

File: data_loader.py (hunk 6 above)
Comment (lsg1213) [reply data_loader.py]: 네 load 시간이 몇초 안 되기도 하고 이렇게 하는 게 더 안전해서요

File: data_loader.py (hunk 7 above)
Comment (lsg1213) [reply data_loader.py]: 넵 변경했습니다

File: data_loader.py (hunk 6 above)
Comment (lsg1213) [reply data_loader.py]: 그렇긴 하네요 딱히 이유는 없습니다 성능에 영향을 주지 않으면 굳이 바꾸지 않아도 괜찮을 것 같은데 영향을 줄까요?

File: data_loader.py (hunk 8 above)
Comment (lsg1213) [reply data_loader.py:210]: 넵 옮겼습니다

File: data_loader.py (hunk 5 above)
Comment (lsg1213) [reply data_loader.py:205]: 추천해주실만한 방법이 있을까요?

File: data_loader.py (hunk 4 above)
Comment (lsg1213) [reply data_loader.py:199]: 수정했습니다

File: data_loader.py (hunk 6 above)
Comment (lsg1213) [reply data_loader.py]: 변경했습니다

File: analyzer.py [hunk 9]
```diff
@@ -183,15 +183,15 @@ def get_ks_test_values(values, perfs, min_samples=1, a=0.05, verbose=False):
...
-        pair['perf']['test_f1score'] = f1
+        # precision = np.squeeze(pair['perf']['val_precision'])
+        # recall = np.squeeze(pair['perf']['val_recall'])
+        # f1 = 2 * precision * recall / (precision + recall + 1e-8)
+        # pair['perf']['val_f1score'] = f1
+
+        # precision = np.squeeze(pair['perf']['test_precision'])
+        # recall = np.squeeze(pair['perf']['test_recall'])
+        # f1 = 2 * precision * recall / (precision + recall + 1e-8)
+        # pair['perf']['test_f1score'] = f1
```
Comment (daniel03c1) [inline analyzer.py]: 이 변경부분 아예 지워서 풀리퀘 하는건 어떤가요? 즉 수정사항 없는 원본으로 돌리는거죠.

File: nas_seldnet.py [hunk 10]
```diff
@@ -18,15 +18,15 @@
 
 args = argparse.ArgumentParser()
 
-args.add_argument('--name', type=str, required=True)
+args.add_argument('--name', type=str, required=True, help='name must be {name}_{divided index}') # 2021_1, 2021_2
 args.add_argument('--train_path', type=str, 
                   default='/datasets/datasets/DCASE2020/foa_tdm_dev')
 args.add_argument('--test_path', type=str, 
                   default='/datasets/datasets/DCASE2020/feat_label')
```
Comment (daniel03c1) [inline nas_seldnet.py]: 2020으로 되어 있는데, 이 폴더가 진혁씨가 알려준 폴더인가요 아니면 2020 예전 버전 코드인가요?

File: analyzer.py (hunk 9 above)
Comment (lsg1213) [reply analyzer.py]: 넵 알겠습니다

File: nas_seldnet.py (hunk 10 above)
Comment (lsg1213) [reply nas_seldnet.py]: 아 이거 제 루트로 했다가 다시 upstream 껄로 롤백하고 풀리 올렸는데 롤백한 루트도 틀린거였네요 고쳐서 올릴게요

File: nas_seldnet.py [hunk 11]
```diff
@@ -214,16 +215,16 @@ def train_and_eval(train_config,
     input_shape = [300, 64, 7]
 
     # TRAIN DATASET
-    train_x = joblib.load(os.path.join(train_config.train_path, '5_5_x.joblib'))
+    train_x = joblib.load(os.path.join(train_config.train_path, 'x.joblib'))
     train_x = np.concatenate(train_x, axis=0)
-    train_y = joblib.load(os.path.join(train_config.train_path, '5_5_y.joblib'))
+    train_y = joblib.load(os.path.join(train_config.train_path, 'y.joblib'))
```
Comment (daniel03c1) [inline nas_seldnet.py:220]: load_seldnet_data 이런 함수 사용 안하고 x, y 형태로 미리 joblib로 모아둔 데이터 만든 다음에 학습하시는건가요?

Comment (lsg1213) [reply nas_seldnet.py:220]: 원래 tdm 1개 1초부터 5개 5초까지 순차적으로 불러와서 할 생각이었는데 일단 tdm 빼고 하느라고 저렇게 했긴  했습니다
